src/unity/toolkits/object_detection/one_shot_object_detection/util/superposition.cpp

```diff
@@ -0,0 +1,128 @@
+/* Copyright © 2019 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+#include <boost/gil/extension/numeric/sampler.hpp>
+#include <boost/gil/extension/numeric/resample.hpp>
+
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/mapping_function.hpp>
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/superposition.hpp>
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/quadrilateral_geometry.hpp>
+
+#define BLACK boost::gil::rgb8_pixel_t(0,0,0)
+#define WHITE boost::gil::rgb8_pixel_t(255,255,255)
+#define RGBA_WHITE boost::gil::rgba8_pixel_t(255,255,255,0)
+
+namespace turi {
+namespace one_shot_object_detection {
+
+namespace data_augmentation {
+
+namespace superposition {
+
+void superimpose_image_rgb(const boost::gil::rgb8_image_t::view_t &masked,
+                           const boost::gil::rgb8_image_t::view_t &mask,
+                           const boost::gil::rgb8_image_t::view_t &transformed,
+                           const boost::gil::rgb8_image_t::view_t &mask_complement,
+                           const boost::gil::rgb8_image_t::view_t &background) {
+  for (int y = 0; y < masked.height(); ++y) {
+    auto masked_row_it = masked.row_begin(y);
+    auto mask_row_it = mask.row_begin(y);
+    auto transformed_row_it = transformed.row_begin(y);
+    auto mask_complement_row_it = mask_complement.row_begin(y);
+    auto background_row_it = background.row_begin(y);
+    for (int x = 0; x < masked.width(); ++x) {
```

Comment:
I think we should really avoid trying to iterate through the pixels of the image manually. It's both error-prone, and is unlikely to be as efficient as the code that Boost GIL provides. As we discovered / discussed offline, let's see if we do something more like
https://stackoverflow.com/questions/47622796/how-to-combine-images-with-boost-gil

Reply:
@nickjong I'll bring this in the PR for #1859 since that will get rid of the mask entirely so we'll have a cleaner implementation that won't have any masks (hence, a more common code path for RGB and RGBA images) and will also use `transform_pixels`.